File: node/ts/test/chat/AuthAccountsServiceTest.ts
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { config, expect, use } from 'chai';
import chaiAsPromised from 'chai-as-promised';

import * as Native from '../../Native.js';
import * as NativeNice from '../../NativeNice.js';
import * as util from '../util.js';
import {
  AuthAccountsService,
  MFA_KEY_NAME_MAX_LENGTH,
  TokioAsyncContext,
} from '../../net.js';
import { SvrKey } from '../../AccountKeys.js';
import { connectAuth, defineTestGrpcCases } from './ServiceTestUtils.js';
import { ErrorCode, LibSignalErrorBase } from '../../Errors.js';

use(chaiAsPromised);

util.initLogger();
config.truncateThreshold = 0;

describe('AuthAccountsService', () => {
  describe('deleteAccount', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_DeleteAccountTests(),
      connectAuth<AuthAccountsService>,
      async (chat: AuthAccountsService, _req: void, _resp: void) => {
        await chat.deleteAccount();
      }
    );
  });

  describe('setRegistrationLock', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_SetRegistrationLockTests(),
      connectAuth<AuthAccountsService>,
      async (
        chat: AuthAccountsService,
        svrKey: Uint8Array<ArrayBuffer>,
        _resp: void
      ) => {
        await chat.setRegistrationLock({ svrKey: new SvrKey(svrKey) });
      }
    );
  });

  describe('clearRegistrationLock', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_ClearRegistrationLockTests(),
      connectAuth<AuthAccountsService>,
      async (chat: AuthAccountsService, _req: void, _resp: void) => {
        await chat.clearRegistrationLock();
      }
    );
  });

  describe('setRegistrationRecoveryPassword', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_SetRegistrationRecoveryPasswordTests(),
      connectAuth<AuthAccountsService>,
      async (
        chat: AuthAccountsService,
        svrKey: Uint8Array<ArrayBuffer>,
        _resp: void
      ) => {
        await chat.setRegistrationRecoveryPassword({
          svrKey: new SvrKey(svrKey),
        });
      }
    );
  });

  describe('setDiscoverableByPhoneNumber', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_SetDiscoverableByPhoneNumberTests(),
      connectAuth<AuthAccountsService>,
      async (chat: AuthAccountsService, discoverable: boolean, _resp: void) => {
        await chat.setDiscoverableByPhoneNumber({ discoverable });
      }
    );
  });

  describe('generateTotpKey', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_GenerateTotpKeyTests(),
      connectAuth<AuthAccountsService>,
      async (
        chat: AuthAccountsService,
        _args: void,
        resp: NativeNice.GenerateTotpKeyOut
      ) => {
        const out = chat.generateTotpKey();
        switch (resp) {
          case 'tooManyTotpKeys':
            await expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({
                code: ErrorCode.TooManyTotpKeys,
              });
            break;
          case 'tooManyMfaKeys':
            await expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({
                code: ErrorCode.TooManyMfaKeys,
              });
            break;
          default:
            expect(await out).to.deep.equal(resp.success);
        }
      }
    );
  });

  describe('confirmTotpKey', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_ConfirmTotpKeyTests(),
      connectAuth<AuthAccountsService>,
      async (
        chat: AuthAccountsService,
        {
          oneTimePassword,
          name,
          createdAt,
          svrKey,
        }: NativeNice.ConfirmTotpKeyArgs,
        resp: NativeNice.ConfirmTotpKeyOut
      ) => {
        Native.TESTING_EnableDeterministicRngForTesting();
        const out = chat.confirmTotpKey({
          oneTimePassword,
          metadata: { name, createdAt },
          svrKey: new SvrKey(svrKey),
          rng: { __deterministicRngSeedForTesting: 0 },
        });
        switch (resp) {
          case 'oneTimePasswordNotVerified':
            await expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({
                code: ErrorCode.MfaNotVerified,
              });
            break;
          case 'tooManyMfaKeys':
            await expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({
                code: ErrorCode.TooManyMfaKeys,
              });
            break;
          default:
            expect(await out).to.equal(resp.success);
        }
      }
    );
  });

  describe('startWebAuthnRegistration', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_StartWebAuthnRegistrationTests(),
      connectAuth<AuthAccountsService>,
      async (
        chat: AuthAccountsService,
        _args: void,
        resp: NativeNice.StartWebAuthnRegistrationOut
      ) => {
        const out = chat.startWebAuthnRegistration();
        switch (resp) {
          case 'tooManyMfaKeys':
            await expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({
                code: ErrorCode.TooManyMfaKeys,
              });
            break;
          default:
            expect(await out).to.deep.equal(resp.success);
        }
      }
    );
  });

  describe('finishWebAuthnRegistration', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_FinishWebAuthnRegistrationTests(),
      connectAuth<AuthAccountsService>,
      async (
        chat: AuthAccountsService,
        {
          attestationObject,
          collectedClientDataJson,
          name,
          createdAt,
          svrKey,
        }: NativeNice.FinishWebAuthnRegistrationArgs,
        resp: NativeNice.FinishWebAuthnRegistrationOut
      ) => {
        Native.TESTING_EnableDeterministicRngForTesting();
        const out = chat.finishWebAuthnRegistration({
          attestationObject,
          collectedClientDataJson,
          metadata: { name, createdAt },
          svrKey: new SvrKey(svrKey),
          rng: { __deterministicRngSeedForTesting: 0 },
        });
        switch (resp) {
          case 'webAuthnRegistrationUnsuccessful':
            await expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({
                code: ErrorCode.WebAuthnRegistrationUnsuccessful,
              });
            break;
          case 'tooManyMfaKeys':
            await expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({
                code: ErrorCode.TooManyMfaKeys,
              });
            break;
          default:
            expect(await out).to.equal(resp.success);
        }
      }
    );
  });

  describe('listMfaKeys', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_ListMfaKeysTests(),
      connectAuth<AuthAccountsService>,
      async (
        chat: AuthAccountsService,
        { svrKey }: NativeNice.ListMfaKeysArgs,
        resp: NativeNice.ListMfaKeysOut
      ) => {
        const out = await chat.listMfaKeys({ svrKey: new SvrKey(svrKey) });
        expect(out).to.deep.equal(
          resp.success.map(({ id, metadata, kind }) => ({
            id,
            metadata: metadata === 'unreadable' ? null : metadata.metadata,
            kind,
          }))
        );
      }
    );
  });

  describe('setMfaKeyMetadata', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_SetMfaKeyMetadataTests(),
      connectAuth<AuthAccountsService>,
      async (
        chat: AuthAccountsService,
        { keyId, name, createdAt, svrKey }: NativeNice.SetMfaKeyMetadataArgs,
        resp: NativeNice.SetMfaKeyMetadataOut
      ) => {
        Native.TESTING_EnableDeterministicRngForTesting();
        const out = chat.setMfaKeyMetadata({
          keyId,
          metadata: { name, createdAt },
          svrKey: new SvrKey(svrKey),
          rng: { __deterministicRngSeedForTesting: 0 },
        });
        switch (resp) {
          case 'success':
            await out;
            break;
          case 'keyNotFound':
            await expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({
                code: ErrorCode.MfaKeyNotFound,
              });
            break;
          default:
            resp satisfies never;
        }
      }
    );
  });

  describe('argument validation', () => {
    const svrKey = new SvrKey(new Uint8Array(32));
    for (const [description, name] of [
      ['too-long', 'a'.repeat(MFA_KEY_NAME_MAX_LENGTH + 1)],
      ['NUL-containing', 'before\0after'],
    ] as const) {
      it(`rejects a ${description} name`, async () => {
        const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
        const [chat] = connectAuth<AuthAccountsService>(tokio);
        const metadata = { name, createdAt: 0 };
        await expect(
          chat.confirmTotpKey({
            oneTimePassword: 123456,
            metadata,
            svrKey,
          })
        )
          .to.eventually.be.rejectedWith(LibSignalErrorBase)
          .and.deep.include({ code: ErrorCode.Generic });
        await expect(chat.setMfaKeyMetadata({ keyId: 0, metadata, svrKey }))
          .to.eventually.be.rejectedWith(LibSignalErrorBase)
          .and.deep.include({ code: ErrorCode.Generic });
        await expect(
          chat.finishWebAuthnRegistration({
            attestationObject: new Uint8Array(0),
            collectedClientDataJson: '{}',
            metadata,
            svrKey,
          })
        )
          .to.eventually.be.rejectedWith(LibSignalErrorBase)
          .and.deep.include({ code: ErrorCode.Generic });
      });
    }

    for (const [description, createdAt] of [
      ['a negative', -1],
      ['a NaN', Number.NaN],
      ['an infinite', Number.POSITIVE_INFINITY],
      ['a fractional', 0.5],
      ['an unsafe', Number.MAX_SAFE_INTEGER + 1],
    ] as const) {
      it(`rejects ${description} creation timestamp from every metadata API`, async () => {
        const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
        const [chat] = connectAuth<AuthAccountsService>(tokio);
        const metadata = { name: 'Work laptop', createdAt };

        await expect(
          chat.confirmTotpKey({
            oneTimePassword: 123456,
            metadata,
            svrKey,
          })
        )
          .to.eventually.be.rejectedWith(LibSignalErrorBase)
          .and.deep.include({ code: ErrorCode.Generic });
        await expect(chat.setMfaKeyMetadata({ keyId: 0, metadata, svrKey }))
          .to.eventually.be.rejectedWith(LibSignalErrorBase)
          .and.deep.include({ code: ErrorCode.Generic });
        await expect(
          chat.finishWebAuthnRegistration({
            attestationObject: new Uint8Array(0),
            collectedClientDataJson: '{}',
            metadata,
            svrKey,
          })
        )
          .to.eventually.be.rejectedWith(LibSignalErrorBase)
          .and.deep.include({ code: ErrorCode.Generic });
      });
    }

    it('rejects an out-of-range key ID', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat] = connectAuth<AuthAccountsService>(tokio);
      // -1 is below the range; 128 is the first ID past it.
      for (const keyId of [-1, 128]) {
        await expect(chat.removeMfaKey({ keyId }))
          .to.eventually.be.rejectedWith(LibSignalErrorBase)
          .and.deep.include({ code: ErrorCode.Generic });
      }
    });
  });

  describe('removeMfaKey', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_RemoveMfaKeyTests(),
      connectAuth<AuthAccountsService>,
      async (
        chat: AuthAccountsService,
        { keyId }: NativeNice.RemoveMfaKeyArgs,
        resp: NativeNice.RemoveMfaKeyOut
      ) => {
        const out = chat.removeMfaKey({ keyId });
        switch (resp) {
          case 'success':
            await out;
            break;
          default:
            resp satisfies never;
        }
      }
    );
  });

  describe('startMfaVerification', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_StartMfaVerificationTests(),
      connectAuth<AuthAccountsService>,
      async (
        chat: AuthAccountsService,
        _noArgs,
        resp: NativeNice.StartMfaVerificationOut
      ) => {
        const out = chat.startMfaVerification({});
        switch (resp) {
          case 'malformed':
            await expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({
                code: ErrorCode.IoError,
              });
            break;
          default: {
            const result = await out;
            expect(result.hasTotp).equals(resp.success.hasTotp);
            if (resp.success.webauthnParams !== null) {
              expect(result.webauthnParams).deep.equals(
                resp.success.webauthnParams
              );
            } else {
              expect(result).does.not.haveOwnProperty('webauthnParams');
            }
          }
        }
      }
    );
  });

  describe('finishMfaVerification', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_FinishMfaVerificationTests(),
      connectAuth<AuthAccountsService>,
      async (
        chat: AuthAccountsService,
        args: NativeNice.BridgeMfaVerificationCredential,
        resp: NativeNice.FinishMfaVerificationOut
      ) => {
        const credential =
          'totp' in args
            ? { totpPassword: args.totp }
            : { webauthnJson: args.webAuthn };
        const out = chat.finishMfaVerification({ credential });
        switch (resp) {
          case 'failedToVerify':
            await expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({
                code: ErrorCode.MfaNotVerified,
              });
            break;
          default:
            await out;
        }
      }
    );
  });
});

File: java/client/src/test/java/org/signal/libsignal/net/AuthAccountsServiceTest.kt
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.net

import kotlinx.coroutines.test.runTest
import org.signal.libsignal.internal.BridgeConfirmedMfaKeyMetadata
import org.signal.libsignal.internal.BridgeMfaKeyKind
import org.signal.libsignal.internal.ConfirmTotpKeyOut
import org.signal.libsignal.internal.GenerateTotpKeyOut
import org.signal.libsignal.internal.ListMfaKeysOut
import org.signal.libsignal.internal.NativeTesting
import org.signal.libsignal.internal.NativeTestingNice
import org.signal.libsignal.internal.RemoveMfaKeyOut
import org.signal.libsignal.internal.SetMfaKeyMetadataOut
import org.signal.libsignal.internal.TokioAsyncContext
import org.signal.libsignal.internal.await
import org.signal.libsignal.net.assertNonSuccess
import java.time.Instant
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertNull

class AuthAccountsServiceTest {
  @Test
  fun testDeleteAccount() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_DeleteAccountTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthAccountsService,
        invoke = { chat, _ ->
          chat.deleteAccount()
        },
        check = { _, actual ->
          assertIs<RequestResult.Success<Unit>>(actual)
        },
      )
    }

  @Test
  fun testSetRegistrationLock() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_SetRegistrationLockTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthAccountsService,
        invoke = { chat, req ->
          chat.setRegistrationLock(svrKey = SvrKey(req))
        },
        check = { _, actual ->
          assertIs<RequestResult.Success<Unit>>(actual)
        },
      )
    }

  @Test
  fun testClearRegistrationLock() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_ClearRegistrationLockTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthAccountsService,
        invoke = { chat, _ ->
          chat.clearRegistrationLock()
        },
        check = { _, actual ->
          assertIs<RequestResult.Success<Unit>>(actual)
        },
      )
    }

  @Test
  fun testSetRegistrationRecoveryPassword() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_SetRegistrationRecoveryPasswordTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthAccountsService,
        invoke = { chat, req ->
          chat.setRegistrationRecoveryPassword(svrKey = SvrKey(req))
        },
        check = { _, actual ->
          assertIs<RequestResult.Success<Unit>>(actual)
        },
      )
    }

  @Test
  fun testSetDiscoverableByPhoneNumber() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_SetDiscoverableByPhoneNumberTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthAccountsService,
        invoke = { chat, req ->
          chat.setDiscoverableByPhoneNumber(discoverable = req)
        },
        check = { _, actual ->
          assertIs<RequestResult.Success<Unit>>(actual)
        },
      )
    }

  @Test
  fun testGenerateTotpKey() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_GenerateTotpKeyTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthAccountsService,
        invoke = { chat, _ ->
          chat.generateTotpKey()
        },
        check = { expected, actual ->
          when (expected) {
            is GenerateTotpKeyOut.Success ->
              assertEquals(
                PendingTotpKey.fromInternal(expected._0),
                assertIs<RequestResult.Success<PendingTotpKey>>(actual).result,
              )
            GenerateTotpKeyOut.TooManyTotpKeys ->
              actual.assertNonSuccess<_, _, TooManyTotpKeysException>()
            GenerateTotpKeyOut.TooManyMfaKeys ->
              actual.assertNonSuccess<_, _, TooManyMfaKeysException>()
          }
        },
      )
    }

  @Test
  fun testConfirmTotpKey() =
    runTest {
      NativeTesting.TESTING_EnableDeterministicRngForTesting()
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_ConfirmTotpKeyTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthAccountsService,
        invoke = { chat, req ->
          chat.confirmTotpKey(
            oneTimePassword = req.oneTimePassword,
            metadata = MfaMetadata(name = req.name, createdAt = req.createdAt),
            svrKey = SvrKey(req.svrKey),
            rngSeedForTesting = DeterministicRandomSeedUseOnlyForTesting(0),
          )
        },
        check = { expected, actual ->
          when (expected) {
            is ConfirmTotpKeyOut.Success ->
              assertEquals(
                expected._0,
                assertIs<RequestResult.Success<Int>>(actual).result,
              )
            ConfirmTotpKeyOut.OneTimePasswordNotVerified ->
              actual.assertNonSuccess<_, _, OneTimePasswordNotVerifiedException>()
            ConfirmTotpKeyOut.TooManyMfaKeys ->
              actual.assertNonSuccess<_, _, TooManyMfaKeysException>()
          }
        },
      )
    }

  @Test
  fun testTotpKeyNameInvalid() =
    runTest {
      val tokioAsyncContext = TokioAsyncContext()
      val (chat, _) = AuthenticatedChatConnection.fakeConnect(tokioAsyncContext, NoOpListener())
      val service = AuthAccountsService(chat)

      for (invalidName in listOf("a".repeat(MfaMetadata.NAME_MAX_LENGTH + 1), "before\u0000after")) {
        val metadata = MfaMetadata(name = invalidName, createdAt = Instant.EPOCH)
        val confirmResult =
          service
            .confirmTotpKey(
              oneTimePassword = 123456,
              metadata = metadata,
              svrKey = SvrKey(ByteArray(32)),
            ).await()
        assertIs<IllegalArgumentException>(assertIs<RequestResult.ApplicationError>(confirmResult).cause)

        val setResult =
          service
            .setMfaKeyMetadata(
              keyId = 0,
              metadata = metadata,
              svrKey = SvrKey(ByteArray(32)),
            ).await()
        assertIs<IllegalArgumentException>(assertIs<RequestResult.ApplicationError>(setResult).cause)
      }
    }

  @Test
  fun testTotpKeyCreatedAtBeforeEpoch() =
    runTest {
      val tokioAsyncContext = TokioAsyncContext()
      val (chat, _) = AuthenticatedChatConnection.fakeConnect(tokioAsyncContext, NoOpListener())
      val service = AuthAccountsService(chat)
      val svrKey = SvrKey(ByteArray(32))
      // There is no upper bound, but the bridge represents timestamps as non-negative milliseconds.
      val unsupportedCreatedAtValues =
        listOf(
          Instant.EPOCH.minusNanos(1),
          Instant.EPOCH.minusMillis(1),
          Instant.ofEpochSecond(-1),
        )

      for (createdAt in unsupportedCreatedAtValues) {
        val metadata = MfaMetadata(name = "test", createdAt = createdAt)

        val confirmResult =
          service
            .confirmTotpKey(
              oneTimePassword = 123456,
              metadata = metadata,
              svrKey = svrKey,
            ).await()
        assertIs<IllegalArgumentException>(
          assertIs<RequestResult.ApplicationError>(confirmResult).cause,
          "confirm for $createdAt",
        )

        val setResult = service.setMfaKeyMetadata(keyId = 0, metadata = metadata, svrKey = svrKey).await()
        assertIs<IllegalArgumentException>(
          assertIs<RequestResult.ApplicationError>(setResult).cause,
          "set for $createdAt",
        )
      }
    }

  @Test
  fun testMfaKeyIdOutOfRange() =
    runTest {
      val tokioAsyncContext = TokioAsyncContext()
      val (chat, _) = AuthenticatedChatConnection.fakeConnect(tokioAsyncContext, NoOpListener())
      val service = AuthAccountsService(chat)

      // -1 is below the range; 128 is the first ID past it.
      for (badId in listOf(-1, 128)) {
        val removeResult = service.removeMfaKey(keyId = badId).await()
        assertIs<IllegalArgumentException>(assertIs<RequestResult.ApplicationError>(removeResult).cause, "for $badId")
      }
    }

  @Test
  fun testListMfaKeys() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_ListMfaKeysTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthAccountsService,
        invoke = { chat, req ->
          chat.listMfaKeys(svrKey = SvrKey(req.svrKey))
        },
        check = { expected, actual ->
          when (expected) {
            is ListMfaKeysOut.Success -> {
              val keys = assertIs<RequestResult.Success<List<ConfirmedMfaKey>>>(actual).result
              assertEquals(expected._0.map { ConfirmedMfaKey.fromInternal(it) }, keys)
              for ((expectedKey, key) in expected._0.zip(keys)) {
                if (expectedKey.metadata is BridgeConfirmedMfaKeyMetadata.Unreadable) {
                  assertNull(key.metadata)
                }
                when (expectedKey.kind) {
                  BridgeMfaKeyKind.Totp -> assertEquals(MfaKeyKind.TOTP, key.kind)
                  BridgeMfaKeyKind.Unknown -> assertEquals(MfaKeyKind.UNKNOWN, key.kind)
                }
              }
            }
          }
        },
      )
    }

  @Test
  fun testSetMfaKeyMetadata() =
    runTest {
      NativeTesting.TESTING_EnableDeterministicRngForTesting()
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_SetMfaKeyMetadataTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthAccountsService,
        invoke = { chat, req ->
          chat.setMfaKeyMetadata(
            keyId = req.keyId,
            metadata = MfaMetadata(name = req.name, createdAt = req.createdAt),
            svrKey = SvrKey(req.svrKey),
            rngSeedForTesting = DeterministicRandomSeedUseOnlyForTesting(0),
          )
        },
        check = { expected, actual ->
          when (expected) {
            SetMfaKeyMetadataOut.Success -> assertIs<RequestResult.Success<Unit>>(actual)
            SetMfaKeyMetadataOut.KeyNotFound ->
              actual.assertNonSuccess<_, _, MfaKeyNotFoundException>()
          }
        },
      )
    }

  @Test
  fun testRemoveMfaKey() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_RemoveMfaKeyTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthAccountsService,
        invoke = { chat, req ->
          chat.removeMfaKey(keyId = req.keyId)
        },
        check = { expected, actual ->
          when (expected) {
            RemoveMfaKeyOut.Success -> assertIs<RequestResult.Success<Unit>>(actual)
          }
        },
      )
    }
}
